### PR TITLE
[Function Contract] Replace Call Mode - Phase 2

### PR DESCRIPTION
**Pull request by LukeW1999**
## Summary

This PR implements precise havoc control through `__ESBMC_assigns` clause and enhances contract functionality with improved support for `__ESBMC_is_fresh`, `__ESBMC_old`, and `__ESBMC_return_value`. Multiple edge cases are fixed and comprehensive test coverage is added.

### `__ESBMC_assigns` Clause

Expression-based assigns clause for precise havoc generation in replace-call mode:

- Only specified memory locations are havoc'd
- Supports global variables, pointer dereferences, array elements
- Supports `__ESBMC_assigns(0)` for pure functions
- Falls back to conservative global havoc if no assigns clause provided (backward compatible)

Example:
```c
void modify_x(int *ptr) {
    __ESBMC_assigns(global_x, *ptr);
    __ESBMC_ensures(global_x == 42);
    __ESBMC_ensures(*ptr == 100);
    global_x = 42;
    *ptr = 100;
}
```

### `__ESBMC_is_fresh` Support

Support for `__ESBMC_is_fresh` in requires and ensures clauses for memory allocation verification.

### Enhanced `__ESBMC_old` Support

- Fixed segfault with multiple `__ESBMC_old` references to same expression
- Improved handling in complex logical expressions (`&&`, `||`)
- Fixed type issues by stripping implicit typecasts
- Fixed handling in conditional assignments from Clang's short-circuit evaluation

### `__ESBMC_return_value` Improvements

- Automatic type resolution based on function return type
- Better support for struct return types
- Fixed type mismatches in ensures clauses

### New Option: `--assume-nonnull-valid`

Assumes non-null pointers are valid, improving verification efficiency for certain cases.

## Bug Fixes

1. **Fixed crash when deleting multiple consecutive ensures ASSUME instructions**
   - Collect all ensures to remove first, find replacement targets that skip all ensures, update GOTO targets, then delete

2. **Fixed `inline_temporary_variables` for conditional assignments**
   - Added `find_all_assignments()` to find all assignments and select non-trivial ones (not constant 0/1)

3. **Fixed array handling in assigns clause**
   - Added `unwrap_array_decay()` to convert `&arr[0]` back to `arr`

4. **Fixed array havoc in replace-call mode**
   - Use `ARRAY_OF(NONDET)` for array types

5. **Fixed `assigns_target` crash in `--enforce-contract` mode**
   - Added bounds checking and error handling

6. **Fixed bounds check in `sym_name_to_symbol`**
   - Added bounds checking before accessing string positions

Additional fixes: type mismatches, instruction ordering, contract intrinsics definition, Cppcheck warnings.

## Testing

Added regression tests covering:
- `__ESBMC_assigns` with various expression types
- `__ESBMC_is_fresh` in requires and ensures clauses
- Multiple `__ESBMC_old` references
- `__ESBMC_old` with logical operators
- Struct return values
- Empty assigns clause
- Array handling
- Conditional assignments
- KNOWNBUG tests for documented limitations

Now 90 function contract regression tests in total with 2 knownbug cases.

## Breaking Changes

None. Backward compatible:
- Functions without `__ESBMC_assigns` clause continue to use conservative global havoc
- All existing contract functionality remains unchanged

## Files Changed

- `src/goto-programs/contracts/contracts.cpp` - Main contract implementation
- `src/goto-programs/contracts/contracts.h` - Contract interface
- `src/goto-programs/builtin_functions.cpp` - Builtin function handling
- `src/clang-c-frontend/clang_c_language.cpp` - Contract intrinsics definition
- `src/esbmc/options.cpp` - New `--assume-nonnull-valid` option
- Regression tests in `regression/function_contract/`

### 1. Dynamic Frame Conditions

Current implementation uses static frame conditions where all locations in `__ESBMC_assigns` are havoc'd, even if only a subset is actually modified.

Impact:
- Functions assigning entire arrays but only modifying specific elements will over-approximate
- Functions depending on global array state modified by previous calls cannot be verified in isolation

Future work: Requires dynamic frame conditions to track actually-modified locations.

Test coverage: Marked as KNOWNBUG tests (`replace_call_dynamic_frame_knownbug`, `replace_call_global_array_dependency_knownbug`).

### 2. Dynamic Indexing

Cannot express "modify element determined by runtime value". For example, `__ESBMC_assigns(nodes[from_idx + 1].counter)` where `from_idx` is a runtime value is not supported.

Workaround: List all possible targets or avoid replace-call mode for such functions. But hard to work if the from_idx is really dynamic.

### 3. Pointer Alias Analysis

`__ESBMC_assigns` requires manual specification of all modified locations. No automatic alias inference is performed.

Example:
```c
void update(int *a, int *b) {
    __ESBMC_assigns(*a, *b);  // Must list both even if a == b
    *a = 10; *b = 20;
}
```

### 4. Complex Ensures with `||` and `__ESBMC_old`

Very complex patterns with `||` and multiple `__ESBMC_old` references may lose conditions during `inline_temporary_variables` processing in some edge cases.

It would might happen when using ternary opeartion with `__ESBMC_old`

Workaround: Split into separate clauses.

## Previously Resolved Issues

The following issues were previously limitations but have been resolved:

- **Array decay in assigns clause**: Added `unwrap_array_decay()` to handle `&arr[0]` → `arr` conversion
- **Array havoc in replace-call mode**: Use `ARRAY_OF(NONDET)` for array types
- **Multiple `__ESBMC_old` references**: Fixed snapshot collection and materialization
- **Conditional assignments in ensures**: Added `find_all_assignments()` to handle short-circuit evaluation

### CODE REVIEW

**Comment by LukeW1999** [comment]
Follow-up: Compositional contract verification
 Changes
1. **`replace "*"` now replaces all calls unconditionally** — removed the old leaf/parent classification that silently kept parent function bodies. Semantics now match the contract user wrote.
2. **Fix enforce + `--function` pointer validity** — entry-point pointer params were nondet/invalid, causing spurious ensures failures. Now auto-assumes valid_object for pointer params.
3. **12 new regression tests** — TLV parser (enforce leaf/mid/top + replace all), Election protocol, Ring topology, multi-level call chains. 121/121 pass

**Comment by rafaelsamenezes** [comment]
@LukeW1999 is this ready for review?

**Comment by LukeW1999** [comment]
Hi Dear @rafaelsamenezes , yes it is ready to review. Hope you are doing well recently.

File: src/goto-programs/builtin_functions.cpp [hunk 1]
```diff
@@ -637,13 +647,130 @@ void goto_convertt::do_function_call_symbol(
...
+          "builtin_functions",
+          "__ESBMC_assigns(0) - pure function (no side effects)");
+
+        // Generate a special marker to indicate explicit empty assigns
+        goto_programt::targett t = dest.add_instruction(ASSERT);
+        t->guard = gen_true_expr();
+        t->location = function.location();
+        t->location.comment("contract::assigns_empty");
+        t->location.property("empty assigns marker");
+
```
[the remark this answers is not in the document]
Comment (LukeW1999) [reply src/goto-programs/builtin_functions.cpp:704]: This review is wrong. Need to use assert to avoid the slicing.

File: src/goto-programs/builtin_functions.cpp [hunk 2]
```diff
@@ -637,13 +647,130 @@ void goto_convertt::do_function_call_symbol(
...
+      // Create a temporary symbol expression as LHS
+      // The name doesn't matter - we only care about the sideeffect on RHS
+      irep_idt tmp_name = "assigns_target$tmp$" + std::to_string(i);
+      symbol_exprt tmp_lhs(tmp_name, actual_arg.type());
+
+      // Generate assignment: tmp = assigns_target(expr)
+      code_assignt assignment(tmp_lhs, assigns_expr);
+      assignment.location() = function.location();
+      copy(assignment, ASSIGN, dest);
+    }
```
[the remark this answers is not in the document]
Comment (LukeW1999) [reply src/goto-programs/builtin_functions.cpp]: fixed

File: src/goto-programs/builtin_functions.cpp [hunk 3]
```diff
@@ -637,13 +647,130 @@ void goto_convertt::do_function_call_symbol(
...
+      // Check if it's &(0) which would be address_of(constant 0) - but this is invalid C
+      // So also check for direct 0/NULL passed (shouldn't happen with macro, but be safe)
+      if (
+        first_arg.is_zero() ||
+        (first_arg.id() == "constant" && first_arg.get("value") == "0"))
+      {
+        log_debug(
+          "builtin_functions",
+          "__ESBMC_assigns(0) - pure function (no side effects)");
+
```
[the remark this answers is not in the document]
Comment (LukeW1999) [reply src/goto-programs/builtin_functions.cpp]: Sorry for the history doc. I changed to ()now instead of (0)

**Comment by LukeW1999** [comment]
@rafaelsamenezes , I have responded to Copliot. What should I do now?

File: src/esbmc/esbmc_parseoptions.cpp [hunk 4]
```diff
@@ -2667,24 +2674,45 @@ void esbmc_parseoptionst::print_ileave_points(
...
-  bool has_enforce)
+  bool has_enforce,
+  bool has_enforce_all,
+  bool has_replace_all)
 {
   namespacet ns(context);
   code_contractst contracts(goto_functions, context, ns);
 
+  // Reference to context for use in lambda
+  contextt &ctx = context;
```
Comment (rafaelsamenezes) [inline src/esbmc/esbmc_parseoptions.cpp:2685]: Don't think you need to create a new variable here. Just capture context directly in the lambda

**Review by rafaelsamenezes — CHANGES_REQUESTED** [review]

File: src/esbmc/esbmc_parseoptions.cpp (hunk 4 above)
Comment (LukeW1999) [reply src/esbmc/esbmc_parseoptions.cpp:2685]: Hi @rafaelsamenezes,

Thanks for the feedback! I'd love to capture it directly, but since context is a class member, C++ doesn't allow capturing it by reference in a lambda without first binding it to a local reference (like ctx). :>

An alternative would be capturing `this`, but I went with the ctx alias to make the capture more explicit and specific to this variable. What are your thoughts on this?

**Review by rafaelsamenezes — APPROVED** [review]
